qcengine/programs/qchem.py

```diff
@@ -0,0 +1,254 @@
+"""
+Calls the Q-Chem executable.
+"""
+
+import os
+from typing import Any, Dict, List, Optional, Tuple
+
+import numpy as np
+
+from qcelemental.models import Result
+from qcelemental.util import parse_version, safe_version, which
+
+from ..exceptions import InputError, UnknownError
+from ..util import disk_files, execute, popen, temporary_directory
+from .model import ProgramHarness
+
+
+class QChemHarness(ProgramHarness):
+    _defaults: Dict[str, Any] = {
+        "name": "QChem",
+        "scratch": True,
+        "thread_safe": False,
+        "thread_parallel": True,
+        "node_parallel": False,
+        "managed_memory": True,
+    }
+    version_cache: Dict[str, str] = {}
+
+    def found(self, raise_error: bool = False) -> bool:
+        return which('qchem',
+                     return_bool=True,
+                     raise_error=raise_error,
+                     raise_msg='Please install by visiting the Q-Chem website.')
+
+    def _get_qc_path(self, config: Optional['JobConfig'] = None):
+        paths = os.environ.copy()
+        paths["QCSCRATCH"] = "/tmp"
+        if config and config.scratch_directory:
+            paths["QCSCRATCH"] = config.scratch_directory
+
+        # Nothing else to pass in
+        if "QC" in os.environ:
+            return paths
+
+        # Assume QC path is
+        qchem_path = which('qchem')
+        if qchem_path:
+            paths["QC"] = os.path.dirname(os.path.dirname(qchem_path))
+
+        return paths
+
+    def get_version(self) -> str:
+        self.found(raise_error=True)
+
+        which_prog = which('qchem')
+        if which_prog not in self.version_cache:
+            with popen([which_prog, '-h'], popen_kwargs={"env": self._get_qc_path()}) as exc:
+                exc["proc"].wait(timeout=15)
+
+            if "QC not defined" in exc["stdout"]:
+                return safe_version("0.0.0")
+
+            self.version_cache[which_prog] = safe_version(exc["stdout"].splitlines()[0].split()[-1])
+
+        return self.version_cache[which_prog]
+
+    def compute(self, input_model: 'ResultInput', config: 'JobConfig') -> 'Result':
+        """
+        Run qchem
+        """
+        # Check if qchem executable is found
+        self.found(raise_error=True)
+
+        # Check qchem version
+        qceng_ver = "5.2"
+        if parse_version(self.get_version()) < parse_version(qceng_ver):
+            raise TypeError(f"Q-Chem version <{qceng_ver} not supported (found version {self.get_version()})")
+            raise TypeError("Q-Chem version '{}' not supported".format(self.get_version()))
+
+        # Setup the job
+        job_inputs = self.build_input(input_model, config)
+
+        # Run qchem
+        exe_success, proc = self.execute(job_inputs)
```

Comment:
* (a) `(exe_success, proc)`
* (b) `(success, dexe)`

I'll standardize on anything, just let me know which, so we don't have two strains of harness running through qcng

Reply:
I like 1) I think.